### PR TITLE
Allow accounts to be created at different times

### PR DESCRIPTION
**Pull request by gakonst**
Closes https://github.com/interledger-rs/interledger-rs/issues/212

It always does exponential backoff now. 

@emschwartz described the following behavior:
> exp backoff, 10 retries ONLY on temporary errors, stops on final errors but returns 202 Accepted if the error was the equivalent of a 404 Not Found)

[RetryIf](https://docs.rs/tokio-retry/0.2.0/tokio_retry/struct.RetryIf.html) might be usable for the above, or although futures-retry probably has a better API: https://gitlab.com/mexus/futures-retry/blob/master/examples/tcp-client.rs. Which status codes do we want to retry and which not?

### CODE REVIEW

**Comment by gakonst** [comment]
> If the problem is either of the first two, the account creation step should probably fail (meaning we can't spawn the call to create the account in all cases).

I agree with this. We should not spawn for these 2 cases.

**Comment by kincaidoneil** [comment]
> To handle the case where the account wasn't created on the other side, it seems more like a change to the SE and the send message API than to the account creation API. The SE needs to understand whatever error code it gets back from trying to send the handshake message as meaning "the account hasn't been created yet" and return a 202 or something when the account is created.

For this case, why does the SE need to know the account wasn't created on the peer's end?

Within SEs, I think this behavior would suffice:

1. On account setup, perform a handshake/attempt to send the message to exchange identifiers (if it fails, that's fine)
2. On sending an outgoing settlement, if no identifier is linked, try to perform the handshake before settling (if it fails again, the settlement amount will be tracked and retried later)
3. As new settlements are triggered (or when the peer sends its handshake message), step 2 can be repeated

(This is more or less the behavior that exists with the JavaScript ETH plugin; it never checks if it's connected to the peer or anything).

**Comment by gakonst** [comment]
From call with Kincaid:

Error retry policy:
- retry timeouts
- retry 5xx
- do not retry 4xx

Account creation process handshake should be:
1. alice tries to add bob to her connector
1. alice's engine tries to send alice's address to bob's engine (via the connector)
1. bob has not configured alice in his connector so the call fails with a terminal error (alice's engine won't retry)
1. account creation moves forward w/o configuring alice's engine
1. bob tries to add alice to his connector
1. bob's engine sends bob's address to alice's engine (via the connector)
1. alice receives that, **and creates bob's account in her store** (this means that bob's data is now fully configured on alice's connector/engine)
1. alice's engine acks and sends bob her address
1. bob's engine saves alice's address
1. alice's account is now fully configured on bob's connector/engine

File: crates/interledger-api/src/client.rs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+use futures_retry::{ErrorHandler, FutureRetry, RetryPolicy};
+use http::StatusCode;
+use log::trace;
+use reqwest::r#async::Client as HTTPClient;
+use serde_json::json;
+use std::fmt::Display;
+use std::time::Duration;
+use url::Url;
+
+// The account creation endpoint set by the engines in [RFC536](https://github.com/interledger/rfcs/pull/536)
```
Comment (emschwartz) [inline crates/interledger-api/src/client.rs]: That isn't the RFC number, that's the PR number

File: crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs [hunk 2]
```diff
@@ -39,15 +41,27 @@ use interledger_settlement::{Convert, ConvertDetails, Quantity};
...
+impl PaymentDetailsRequest {
+    fn new(challenge: Vec<u8>) -> Self {
+        PaymentDetailsRequest { challenge }
+    }
+}
+
 #[derive(Debug, Clone, Serialize, Deserialize)]
 struct PaymentDetailsResponse {
     to: Addresses,
     sig: Signature,
```
Comment (emschwartz) [inline crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs]: nit: any reason not to call it `signature`?

File: crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs [hunk 3]
```diff
@@ -840,13 +863,53 @@ where
...
+                            move |_| {
+                                // ACK BACK
+                                if let Some(challenge) = payment_details.challenge {
+                                    // if we were challenged, we must respond
+                                    let data = prefixed_message(challenge);
+                                    let signature = signer.sign_message(&data);
+                                    let resp = {
+                                        // Respond with our address, a signature,
+                                        // and no challenge, since we already sent
+                                        // them one earlier
```
Comment (emschwartz) [inline crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs]: Does this work even if the first request to do the handshake failed?

File: crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs [hunk 4]
```diff
@@ -864,20 +927,72 @@ where
...
+            debug!("Received payment details: {:?}", resp);
+            let guard = self.challenges.read();
+            let fut = if let Some(challenge) = (*guard).get(&account_id) {
+                // if we sent them a challenge, we will verify the received
+                // signature and address, and if sig verification passes, we'll
+                // save them in our store
+                let data = prefixed_message(challenge.to_vec());
+                let challenge_hash = Sha3::digest(&data);
+                let recovered_address = resp.sig.recover(&challenge_hash);
+                Either::A(
```
Comment (emschwartz) [inline crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs]: This is synchronous code so it seems kind of strange to use futures here

File: crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs [hunk 5]
```diff
@@ -864,20 +927,72 @@ where
...
+            } else {
+                Either::B(ok(()))
+            };
+
+            Box::new(
+                fut.and_then(move |_| Ok((StatusCode::from_u16(200).unwrap(), "OK".to_string()))),
+            )
+        } else {
+            Box::new(err((
+                StatusCode::from_u16(502).unwrap(),
```
Comment (emschwartz) [inline crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs]: Shouldn't this be a 400?

File: crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs [hunk 6]
```diff
@@ -864,20 +930,74 @@ where
...
```
Comment (emschwartz) [inline crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs]:
```suggestion
                StatusCode::from_u16(400).unwrap(),
```

File: crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs [hunk 7]
```diff
@@ -864,20 +930,74 @@ where
...
+                }
+            } else {
+                Either::B(ok(()))
+            };
+
+            Box::new(
+                fut.and_then(move |_| Ok((StatusCode::from_u16(200).unwrap(), "OK".to_string()))),
+            )
+        } else {
+            Box::new(err((
```
Comment (emschwartz) [inline crates/interledger-settlement-engines/src/engines/ethereum_ledger/eth_engine.rs]:
```suggestion
            error!("Ignoring message that was neither a PaymentDetailsRequest nor a PaymentDetailsResponse");
            Box::new(err((
```

**Pull request merged**